src/mbedtls.c

```diff
@@ -208,34 +208,40 @@ _libssh2_mbedtls_bignum_random(_libssh2_bn *bn, int bits, int top, int bottom)
 {
     size_t len;
     int err;
-    int i;
+    int i, msb;
 
     if (!bn || bits <= 0)
         return -1;
 
-    len = (bits + 1) >> 3;
+    len = (bits + 7) >> 3;
     err = mbedtls_mpi_fill_random(bn, len, mbedtls_ctr_drbg_random, &_libssh2_mbedtls_ctr_drbg);
     if (err)
         return -1;
 
-    /* If `top` is -1, the most significant bit of the random number can be zero. */
-    err = mbedtls_mpi_set_bit(bn, 0, 0);
-    if (err)
-        return -1;
+    /* Zero unsued bits above the most significant bit and if `top` is -1,
+       the most significant bit of the random number can be zero. */
+    msb = bits;
+    if (top == -1)
```

Comment:
I don't think this is correct. "Can be zero" just means that it's not forced to 1 as the other values of `top` do.